agent/function/code_retrieve_agent.py

```diff
@@ -0,0 +1,197 @@
+        ```json
+        [{{"path": "...", "type": "...", "size": ...}}, ...]
+        ```
+        """
+
+    def project_prompt(self) -> str:
+        return f"""
+        User's project: {self.project.name}
+        User's requirements: {self.project.requirement}
+        Description of the project: {self.project.description}
+        Project language: {self.project.lang}
+        """
+
+    def gen_keywords(self) -> list[str]:
+        chat_history = [
+            {"role": "system", "content": self.system_pmpt_retrival_keywords()},
+            {"role": "user", "content": self.project_prompt()},
+        ]
+        return ast.literal_eval(self.model.query(chat_history))
+
+    def retrieve_repos(self, all_repos) -> list[dict]:
+        chat_history = [
+            {"role": "system", "content": self.system_pmpt_retrival_repos()},
+            {"role": "user", "content": self.project_prompt()},
+            {"role": "user", "content": f"github repository lists: {all_repos}"},
+        ]
+        match = re.search(
+            r"```json([^`]*)```", self.model.query(chat_history), re.DOTALL
+        )
+        if match:
+            return ast.literal_eval(match.group(1).strip())
+        return None
+
+    def retrieve_codes(self, repo, paths) -> list[dict]:
+        paths = [
+            {
+                "path": path["path"],
+                "type": path["type"],
+                "size": path["size"],
+            }
+            for path in paths
+        ]
+
+        chat_history = [
+            {"role": "system", "content": self.system_pmpt_retrival_codes()},
+            {"role": "user", "content": self.project_prompt()},
+            {"role": "user", "content": f"github repository: {repo}"},
+            {"role": "user", "content": f"repository files: {paths}"},
+        ]
+
+        output = self.model.query(chat_history)
+        match = re.search(r"```json([^`]*)```", output, re.DOTALL)
+        if match:
+            return ast.literal_eval(match.group(1).strip())
+        return None
+
+    def invoke(self):
+        """
+        Invoke the agent.
+        """
+        # step 1: find relevant github repos
+        with self.console.status("Searching relevant repositories from GitHub..."):
+            relevant_repos = []
+            for kw in self.gen_keywords():
+                repos = retrieve_repos(kw, self.project.lang, self.token)
+                relevant_repos.extend(
+                    [
+                        {
+                            "name": repo["full_name"],
+                            "description": repo["description"],
+                            "forks": repo["forks_count"],
+                            "stars": repo["stargazers_count"],
+                            # "license": repo.get("license", {}).get("name", ""),
+                        }
+                        for repo in repos[:5]
+                    ]
+                )
+            relevant_repos = self.retrieve_repos(relevant_repos)
```

Comment:
yeah, as you see, two same function names make the code a little bit hard to read.